### PR TITLE
[client] Race home and foreign relay on connection setup

### PR DESCRIPTION
**Pull request by pappz**
## Describe your changes

Open peer relay connections by racing the home relay and the remote peer's
relay in parallel, using whichever connects first (falls back to the other
after a delay). Foreign relay clients are cached in a dedicated
ForeignRelaysStore (singleflight dedup + refcount) so the cache mutex is
never held during a network connect.

### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [ ] Is a feature enhancement
- [x] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] This change does **not** modify the public API, gRPC protocols, functionality behavior, CLI / service flags, or introduce a new feature — **OR** I have discussed it with the NetBird team beforehand (link the issue / Slack thread in the description). See [CONTRIBUTING.md](https://github.com/netbirdio/netbird/blob/main/CONTRIBUTING.md#discuss-changes-with-the-netbird-team-first).

> By submitting this pull request, you confirm that you have read and agree to the terms of the [Contributor License Agreement](https://github.com/netbirdio/netbird/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT.md).

## Documentation
Select exactly one:

- [ ] I added/updated documentation for this change
- [x] Documentation is **not needed** for this change (explain why)

### Docs PR URL (required if "docs added" is checked)
Paste the PR link from https://github.com/netbirdio/docs here:

https://github.com/netbirdio/docs/pull/__

## Summary by CodeRabbit

* **New Features**
  * Added foreign-relay connection caching to improve reuse and lifecycle management.
  * Introduced a timed fallback mechanism that races an alternate relay path when the preferred route fails or stalls.
* **Bug Fixes**
  * Improved relay target selection and connection cleanup for more reliable peer connections.
  * Improved recovery when relay server IP details are unavailable.
* **Refactor**
  * Unified relay dialing around structured relay server information and consistent connection handling.
* **Tests**
  * Expanded coverage for fallback behavior, relay caching, cleanup, reconnection, and connection lifecycle events.

### CODE REVIEW

File: shared/relay/client/foreign_relays_store.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		f.mu.RUnlock()
+		if ok {
+			return fr, nil
+		}
+
+		relayClient := NewClientWithServerIP(remoteRelayServer.Addr, remoteRelayServer.IP, f.tokenStore, f.peerID, f.mtu)
+		relayClient.SetTransportFallback(f.transportFallback)
+		if err := relayClient.Connect(f.ctx); err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (pappz) [reply shared/relay/client/foreign_relays_store.go:83]: Client.Connect(ctx) ties the ctx to the connection's lifetime, not just the dial (listenForStopEvents → c.close(true) on ctx.Done(), client.go:738). The foreign client is intentionally long-lived (cached, shared, reaped by cleanup), so it must use f.ctx; the per-attempt ctx would tear down the shared connection on race-cancel/timeout. The race-cancellable part — the peer conn — already uses the caller ctx via fr.client.OpenConn(ctx, peerKey), and the dial is bounded by DefaultConnectionTimeout.

**Review by theodorsm — CHANGES_REQUESTED** [review]
This is already way cleaner than how it used to be with relayTracks, which had a few bugs. We have some manager tests that cover the new behavior in this PR, but we should also do some unit testing on `ForeignRelaysStore` and `FallbackOpener`.

A new E2E edge case to consider: what happens if both sides race and different relays wins due to different network paths? Could we add a heuristic to solve this?

**Comment by pappz** [comment]
> This is already way cleaner than how it used to be with relayTracks, which had a few bugs. We have some manager tests that cover the new behavior in this PR, but we should also do some unit testing on `ForeignRelaysStore` and `FallbackOpener`.
> 
> A new E2E edge case to consider: what happens if both sides race and different relays wins due to different network paths? Could we add a heuristic to solve this?

Sure, I will add unit test.

"**If both sides race, different relays might win**." This new race mechanism is just a fallback. In the first phase, both peers just try to connect on only one server. If it is not successful (because one peer cannot reach the chosen server), then they start connecting in parallel on the other server too. But because the other peer cannot reach the original server, there will not be congestion. Anyway, the server drains all connection attempts and unsubscribes from the peer notifications on the losing servers.

But logically, the code does not protect against the situation where both peers open a 1-1 subscription for the remote peer on two servers (so 4 subscriptions). In this case, if all subscriptions report peer-online messages, we will generate 4 net.Conn instances, and the upper layer will simply time out these connections. Right now, the relay library cleans up the net.Conn instances when we get a peer-went-offline message from the server, but the remote peer doesn't necessarily close the server connection if other peers are using the same server for other connections. So the proper solution is that the server should send a peer-went-offline message in cases where the other peer has also sent an unsubscribe message.

So, in a split-brain situation, the peer should explicitly unsubscribe from the losing server, and this action should explicitly close the net.Conn on the remote side. This triggers a reconnection that resolves the situation.

**Comment by pappz** [comment]
> This is already way cleaner than how it used to be with relayTracks, which had a few bugs. We have some manager tests that cover the new behavior in this PR, but we should also do some unit testing on `ForeignRelaysStore` and `FallbackOpener`.
> 
> A new E2E edge case to consider: what happens if both sides race and different relays wins due to different network paths? Could we add a heuristic to solve this?

Unit tests have been added. The connection logic has been discussed verbally.

**Review by theodorsm — APPROVED** [review]
